contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol

```diff
@@ -0,0 +1,424 @@
+                // just need a reasonable upper bound to avoid unnecessarily re-approving.
+                takerFeeToken.approveIfBelow(erc20Proxy, order.takerFee);
+            } else if (takerFeeToken == takerToken){
+                // Taker fee is payable in the taker token, so we need to
+                // reduce the fill amount to cover the fee.
+                // takerTokenFillAmount' =
+                //   (takerTokenFillAmount * order.takerAssetAmount) /
+                //   (order.takerAssetAmount + order.takerFee)
+                takerTokenFillAmount = LibMathV06.getPartialAmountCeil(
+                    order.takerAssetAmount,
+                    order.takerAssetAmount.safeAdd(order.takerFee),
+                    takerTokenFillAmount
+                );
+            } else {
+                //  Only support taker or maker asset denominated taker fees.
+                LibTransformERC20RichErrors.InvalidTakerFeeTokenError(
+                    address(takerFeeToken)
+                ).rrevert();
+            }
+        }
+
+        // Clamp fill amount to order size.
+        takerTokenFillAmount = LibSafeMathV06.min256(
+            takerTokenFillAmount,
+            order.takerAssetAmount
+        );
+
+        // Perform the fill.
+        return _fillOrder(
+            order,
+            signature,
+            takerTokenFillAmount,
+            protocolFee,
+            makerToken,
+            takerFeeToken == takerToken
+        );
+    }
+
+    /// @dev Try to buy up to `buyAmount` from an order.
+    /// @param makerToken The maker/buy token.
+    /// @param takerToken The taker/sell token.
+    /// @param order The order to fill.
+    /// @param signature The signature for `order`.
+    /// @param buyAmount Amount of maker token to buy.
+    /// @param protocolFee The protocol fee needed to fill `order`.
+    function _buyFromOrder(
+        IERC20TokenV06 makerToken,
+        IERC20TokenV06 takerToken,
+        IExchange.Order memory order,
+        bytes memory signature,
+        uint256 buyAmount,
+        uint256 protocolFee
+    )
+        private
+        returns (FillOrderResults memory results)
+    {
+        IERC20TokenV06 takerFeeToken = order.takerFeeAssetData.length == 0
+            ? IERC20TokenV06(address(0))
+            : _getTokenFromERC20AssetData(order.takerFeeAssetData);
+
+        uint256 makerTokenFillAmount = buyAmount;
+
+        if (order.takerFee != 0) {
+            if (takerFeeToken == makerToken) {
+                // Taker fee is payable in the maker token.
+                // Increase the fill amount to account for maker tokens being
+                // lost to the taker fee.
+                // makerTokenFillAmount' =
+                //  (order.makerAssetAmount * makerTokenFillAmount) /
+                //  (order.makerAssetAmount - order.takerFee)
+                makerTokenFillAmount = LibMathV06.getPartialAmountCeil(
```

Comment:
We could reduce the computation and potential for rounding errors by computing the `takerTokenFillAmount` here. I believe it works out to just:
```
takerTokenFillAmount  = LibMathV06.getPartialAmountCeil(
    makerTokenFillAmount,
    order.makerAssetAmount.safeSub(order.takerFee), 
    order.takerAssetAmount
);  
```

Reply:
nice!